communicator/winrm/communicator.go

```diff
@@ -0,0 +1,186 @@
+package winrm
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"time"
+
+	"github.com/hashicorp/terraform/communicator/remote"
+	"github.com/hashicorp/terraform/terraform"
+	"github.com/masterzen/winrm/winrm"
+	"github.com/packer-community/winrmcp/winrmcp"
+)
+
+// Communicator represents the WinRM communicator
+type Communicator struct {
+	connInfo *connectionInfo
+	client   *winrm.Client
+	endpoint *winrm.Endpoint
+}
+
+// New creates a new communicator implementation over WinRM.
+//func New(endpoint *winrm.Endpoint, user string, password string, timeout time.Duration) (*communicator, error) {
+func New(s *terraform.InstanceState) (*Communicator, error) {
+	connInfo, err := parseConnectionInfo(s)
+	if err != nil {
+		return nil, err
+	}
+
+	endpoint := &winrm.Endpoint{
+		Host:     connInfo.Host,
+		Port:     connInfo.Port,
+		HTTPS:    connInfo.HTTPS,
+		Insecure: connInfo.Insecure,
+		CACert:   connInfo.CACert,
+	}
+
+	comm := &Communicator{
+		connInfo: connInfo,
+		endpoint: endpoint,
+	}
+
+	return comm, nil
+}
+
+// Connect implementation of communicator.Communicator interface
+func (c *Communicator) Connect(o terraform.UIOutput) error {
+	if c.client != nil {
+		return nil
+	}
+
+	params := winrm.DefaultParameters()
+	params.Timeout = formatDuration(c.Timeout())
+
+	client, err := winrm.NewClientWithParameters(
+		c.endpoint, c.connInfo.User, c.connInfo.Password, params)
+	if err != nil {
+		return err
+	}
+
+	if o != nil {
+		o.Output(fmt.Sprintf(
+			"Connecting to remote host via WinRM...\n"+
+				"  Host: %s\n"+
+				"  Port: %d\n"+
+				"  User: %s\n"+
+				"  Password: %t\n"+
+				"  HTTPS: %t\n"+
+				"  Insecure: %t\n"+
+				"  CACert: %t",
+			c.connInfo.Host,
+			c.connInfo.Port,
+			c.connInfo.User,
+			c.connInfo.Password != "",
+			c.connInfo.HTTPS,
+			c.connInfo.Insecure,
+			c.connInfo.CACert != nil,
+		))
+	}
+
+	log.Printf("connecting to remote shell using WinRM")
+	shell, err := client.CreateShell()
+	if err != nil {
+		log.Printf("connection error: %s", err)
+		return err
+	}
+
+	err = shell.Close()
```

Comment:
Probably just a lack of domain knowledge here, but why do you create and then close a shell here? Is it part of the initialization of a proper client or does it function more as a smoke test to catch potential problems early?